common/page.go

```diff
@@ -364,6 +363,105 @@ func (p *Page) initEvents() {
 	}()
 }
 
+// urlTagName is used to check the incoming metric url tag against user
+// defined url regexes. When a match is found a user defined name, which is to
+// be used in the urls place in the url metric tag, is returned.
+//
+// The check is done by calling the handlers that were registered with
+// `page.on('metric')`. The user will need to use `Tag` to supply the
+// url regexes and the matching is done from within there. If a match is found,
+// the supplied name is returned back upstream to the caller of urlTagName.
+func (p *Page) urlTagName(ctx context.Context, urlTag string) (string, bool) {
+	p.eventHandlersMu.RLock()
+
+	// If there are no handlers for EventConsoleAPICalled.
+	if _, ok := p.eventHandlers[EventPageMetricCalled]; !ok {
+		p.eventHandlersMu.RUnlock()
+
+		return "", false
+	}
+
+	var name string
+	var nameChanged bool
+	em := &MetricEvent{
+		urlTag: urlTag,
+	}
+
+	for _, h := range p.eventHandlers[EventPageMetricCalled] {
+		// A handler can register another handler from within itself. This is
+		// the reason to unlock the mutex before calling the handler.
+		p.eventHandlersMu.RUnlock()
+
+		// Call and wait for the handler to complete.
+		h(em)
+
+		p.eventHandlersMu.RLock()
+	}
+	p.eventHandlersMu.RUnlock()
+
+	// If a match was found then the name field in em will have been updated.
+	if em.name != nil {
+		name = *em.name
+		nameChanged = true
+	}
+
+	p.logger.Debugf("urlTagName", "name: %q nameChanged: %v", name, nameChanged)
+
+	return name, nameChanged
+}
+
+// MetricEvent is the type that is exported to JS. It is currently only used to
+// match on the urlTag and return a name when a match is found.
+type MetricEvent struct {
+	// The URL value from the metric's url tag. It will be used to match
+	// against the URL grouping regexs.
+	urlTag string
+
+	// When a match is found this name field should be updated.
+	name *string
+}
+
+// URLOverrides will contain all the URL groupings.
+type URLOverrides struct {
+	URLs []URLOverride `js:"urls"`
+}
+
+// URLOverride contains the single url regex and the name to give to the metric
+// if a match is found.
+type URLOverride struct {
+	// This is a regex.
+	URL string
+	// The name to send back to the caller of the handler.
+	Name string
+}
+
+type regexCallback func(pattern, url string) (bool, error)
+
+// Tag will find the first match given the URLOverrides and the URL from
+// the metric tag and update the name field.
+func (e *MetricEvent) Tag(callBack regexCallback, overrides URLOverrides) error {
+	for _, o := range overrides.URLs {
+		name := strings.TrimSpace(o.Name)
+		if name == "" {
+			return fmt.Errorf("name %q is invalid", o.Name)
+		}
+
+		// callback is a function that will perform the regex test in the Sobek
+		// runtime.
+		val, err := callBack(o.URL, e.urlTag)
```

Comment:
`val` or `matched`? What does a value mean in this context? I'd make it `matched` if it's about matching a URL to a user-provided pattern. And, `e.tagName = userProvidedTagName` would make it clearer instead of `e.name = &name`. Using a pointer is also confusing. Maybe add a field like `isUserURLTagNameExist bool` instead?

Reply:
Updated val to matched in e6d3c398589090d56c89349478c7a1fc98c57504

Reply:
Renamed name to userProvidedTagName 826b52711ea6ca3e9522372f3b6cf873bc2e970d

Reply:
Added bool check in a42e4e709031a79471fbc373da09563c3e84264b